blocks/log/sqlite.py

```diff
@@ -0,0 +1,240 @@
+"""SQLite backend for the main loop log."""
+import sqlite3
+import warnings
+from collections import MutableMapping, Mapping
+from operator import itemgetter
+
+import numpy
+import six
+from six.moves import cPickle, map
+
+from blocks.config import config
+from .log import _TrainingLog
+
+
+ANCESTORS_QUERY = """
+WITH parents (parent, child) AS (
+    SELECT uuid, value FROM status
+    WHERE key = 'resumed_from' AND uuid = ?
+    UNION ALL
+    SELECT uuid, value FROM status
+    INNER JOIN parents ON status.uuid = parents.child
+    WHERE key = 'resumed_from'
+),
+ancestors AS (SELECT parent FROM parents)
+"""
+
+LARGE_BLOB_WARNING = """
+
+A {} object of {} bytes was stored in the SQLite database. SQLite natively \
+only supports numbers and text. Other objects will be pickled before being \
+saved. For large objects, this can be slow and degrade performance of the \
+database."""
+
+
+def adapt_obj(obj):
+    blob = sqlite3.Binary(cPickle.dumps(obj))
+    if len(blob) > 1024 * 4:
+        warnings.warn('large objects stored in SQLite' +
+                      LARGE_BLOB_WARNING.format(type(obj), len(blob)))
+        # Prevent the warning with variable message from repeating
+        warnings.filterwarnings('ignore', 'large objects .*')
+    return blob
+
+
+def adapt_ndarray(obj):
+    if obj.ndim == 0:
+        return float(obj)
+    else:
+        return adapt_obj(obj)
+
+
+def _get_row(row, key):
+    """Handle the returned row e.g. unpickle if needed."""
+    if row is not None:
+        value = row[0]
+        # Resumption UUIDs are stored as bytes and should not be unpickled
+        if (isinstance(value, (sqlite3.Binary, bytes)) and
+                key != 'resumed_from'):
+            value = cPickle.loads(bytes(value))
+        return value
+    raise KeyError(key)
+
+
+def _register_adapter(value, key):
+    """Register an adapter if the type of value is unknown."""
+    # Assuming no storage of non-simple types on channel 'resumed_from'
+    if (not isinstance(value, (type(None), int, float, six.string_types,
+                               bytes, numpy.ndarray)) and
+            key != 'resumed_from'):
+        sqlite3.register_adapter(type(value), adapt_obj)
+
+
+class SQLiteLog(_TrainingLog, Mapping):
+    r"""Training log using SQLite as a backend.
+
+    Parameters
+    ----------
+    database : str, optional
+        The database (file) to connect to. Can also be `:memory:`. See
+        :func:`sqlite3.connect` for details. Uses `config.sqlite_database`
+        by default.
+    \*\*kwargs
+        Arguments to pass to :class:`_TrainingLog`
```

Comment:
If we're going to refer to `_TrainingLog` in the docs should we give it a better name, like `TrainingLogMixin`?